### PR TITLE
Add iterable supports for v-for 

### PR DESCRIPTION
**Pull request by mymyoux**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

Adds support for iterable object implementing [Symbol.iterator]()
It shouldn't have an impact for existing code because non iterable objects are not impacted.
Vue will have the same behaviour than a for...of expression for iterables. (actual interest for using iterables)
Allows us to pass an iterable directly to a template without giving internal properties or converting it to an array

I wrote 4 new tests. No existing tests needed changes.
All tests passed. I ran tests with PhantomJS (lack of Symbol support), Firefox and Chrome.

### CODE REVIEW

**Comment by mitar** [comment]
Related: #5893

**Comment by mitar** [comment]
So I have been using the following for a while now: https://github.com/vuejs/vue/blob/cb062c50cdd21628dcf758fe54c84d17d75cb2ca/src/core/instance/render-helpers/render-list.js

But I thought that in #5893 it is said that this will not go into core for now?

File: src/core/instance/render-helpers/render-list.js [hunk 1]
```diff
@@ -25,11 +25,27 @@ export function renderList (
...
+    if (hasSymbol && val[Symbol.iterator]) {
+      const iterator:Iterator<any> = val[Symbol.iterator]()
+      ret = []
+      i = 0
+      let next
+      do {
+        next = iterator.next()
+        if (!next.done) {
+          key = i
+          ret.push(render(next.value, key, i))
```
Comment (mitar) [inline src/core/instance/render-helpers/render-list.js]: Why not just `ret.push(render(next.value, ret.length))`?

Comment (mymyoux) [reply src/core/instance/render-helpers/render-list.js]: good point I can update this

**Comment by mymyoux** [comment]
> But I thought that in #5893 it is said that this will not go into core for now?

I'm aware of this but I think if you choose to implement [Symbol.iterator] you choose to break old browsers compatibility. This way your iterable object has the same behaviour in Vue than in other parts of your code.
What do you think ?

**Comment by mitar** [comment]
I personally like it, I was all for #5893.

**Pull request merged**